fastly/base_fastly_service_v1.go

```diff
@@ -0,0 +1,494 @@
+package fastly
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"time"
+
+	gofastly "github.com/fastly/go-fastly/fastly"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+)
+
+var fastlyNoServiceFoundErr = errors.New("No matching Fastly Service found")
+
+// ServiceAttributeDefinition provides an interface for service attributes.
+// We compose a service resource out of attribute objects to allow us to construct both the VCL and Wasm service
+// resources from common components.
+type ServiceAttributeDefinition interface {
+	// Register add the attribute to the resource schema.
+	Register(d *schema.Resource) error
+
+	// Read refreshes the attribute state against the Fastly API.
+	Read(d *schema.ResourceData, s *gofastly.ServiceDetail, conn *gofastly.Client) error
+
+	// Process creates or updates the attribute against the Fastly API.
+	Process(d *schema.ResourceData, latestVersion int, conn *gofastly.Client) error
+
+	// HasChange returns whether the state of the attribute has changed against Terraform stored state.
+	HasChange(d *schema.ResourceData) bool
+
+	// MustProcess returns whether we must process the resource (usually HasChange==true but allowing exceptions).
+	// For example: at present, the settings attributeHandler (block_fastly_service_v1_settings.go) must process when
+	// default_ttl==0 and it is the initialVersion - as well as when default_ttl or default_host have changed.
+	MustProcess(d *schema.ResourceData, initialVersion bool) bool
```

Comment:
Just noting that we talked offline about how it is okay that we just use `initialVersion` here for now rather than solving the more general use of this method.